contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,1006 @@
+            ctx->convergenceCounter = 0;
+            DEBUG(2, "convergence counter reset, no change applied\n");
+        }
+        else if (boundChange != 0) {
+            ctx->compressionLevel -= boundChange;
+            ctx->cooldown = CLEVEL_DECREASE_COOLDOWN;
+            ctx->convergenceCounter = 1;
+
+            DEBUG(2, "create or write threads waiting on compression, tried to decrease compression level by %u\n\n", boundChange);
+        }
+    }
+    else if (1-compressWaitWriteCompletion > threshold || 1-compressWaitCreateCompletion > threshold) {
+        /* compress waiting on write */
+        double const completion = MIN(compressWaitWriteCompletion, compressWaitCreateCompletion);
+        unsigned const change = convertCompletionToChange(completion);
+        unsigned const boundChange = MIN(change, ZSTD_maxCLevel() - ctx->compressionLevel);
+        if (ctx->convergenceCounter >= CONVERGENCE_LOWER_BOUND && boundChange != 0) {
+            /* reset convergence counter, might have been a spike */
+            ctx->convergenceCounter = 0;
+            DEBUG(2, "convergence counter reset, no change applied\n");
+        }
+        else if (boundChange != 0) {
+            ctx->compressionLevel += boundChange;
+            ctx->cooldown = 0;
+            ctx->convergenceCounter = 1;
+
+            DEBUG(2, "compress waiting on write or create, tried to increase compression level by %u\n\n", boundChange);
+        }
+
+    }
+
+    if (ctx->compressionLevel == prevCompressionLevel) {
+        ctx->convergenceCounter++;
+    }
+
+    if (g_forceCompressionLevel) {
+        ctx->compressionLevel = g_compressionLevel;
+    }
+}
+
+static size_t getUseableDictSize(unsigned compressionLevel)
+{
+    ZSTD_parameters const params = ZSTD_getParams(compressionLevel, 0, 0);
+    unsigned const overlapLog = compressionLevel >= (unsigned)ZSTD_maxCLevel() ? 0 : 3;
+    size_t const overlapSize = 1 << (params.cParams.windowLog - overlapLog);
+    return overlapSize;
+}
+
+static void* compressionThread(void* arg)
+{
+    adaptCCtx* const ctx = (adaptCCtx*)arg;
+    unsigned currJob = 0;
+    for ( ; ; ) {
+        unsigned const currJobIndex = currJob % ctx->numJobs;
+        jobDescription* const job = &ctx->jobs[currJobIndex];
+        DEBUG(2, "starting compression for job %u\n", currJob);
+
+        {
+            /* check if compression thread will have to wait */
+            unsigned willWaitForCreate = 0;
+            unsigned willWaitForWrite = 0;
+
+            pthread_mutex_lock(&ctx->jobReady_mutex.pMutex);
+            if (currJob + 1 > ctx->jobReadyID) willWaitForCreate = 1;
+            pthread_mutex_unlock(&ctx->jobReady_mutex.pMutex);
+
+            pthread_mutex_lock(&ctx->jobWrite_mutex.pMutex);
+            if (currJob - ctx->jobWriteID >= ctx->numJobs) willWaitForWrite = 1;
+            pthread_mutex_unlock(&ctx->jobWrite_mutex.pMutex);
+
+
+            if (willWaitForCreate) {
```

Comment:
What happens if `willWaitForCreate == 0` ?
It seems that `ctx->compressWaitCreateCompletion` will not be updated,
so it will keep its previous value.
And this value will be used in next `adaptCompressionLevel` decision ...